File: travel-backend-java/src/main/java/com/travel/backend/service/impl/ItineraryEnricher.java
package com.travel.backend.service.impl;

import tools.jackson.databind.json.JsonMapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import tools.jackson.databind.ObjectMapper;
import com.travel.backend.common.ItineraryEventPublisher;
import com.travel.backend.dto.AgentButlerNoteRequest;
import com.travel.backend.dto.AgentButlerNoteResponse;
import com.travel.backend.dto.AgentGenerateResponse;
import com.travel.backend.dto.AgentPoiIntrosRequest;
import com.travel.backend.dto.AgentPoiIntrosResponse;
import com.travel.backend.dto.GenerateRequest;
import com.travel.backend.entity.ItineraryDay;
import com.travel.backend.entity.ItineraryItem;
import com.travel.backend.entity.ItineraryMain;
import com.travel.backend.mapper.ItineraryDayMapper;
import com.travel.backend.mapper.ItineraryItemMapper;
import com.travel.backend.mapper.ItineraryMainMapper;
import com.travel.backend.service.AgentService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.CacheManager;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * 行程生成收尾的辅助写回：备选池落库、AI 管家讲解与景点介绍富化。
 *
 * <p>独立成 Bean 是为了让富化挂到专用有界池（enricherExecutor）上：富化是可选增强，
 * 不能与主生成流程抢占线程资源，也不能沿用公共 ForkJoinPool 造成无界并发。</p>
 */
@Slf4j
@Service
public class ItineraryEnricher {

    private final AgentService agentService;
    private final ItineraryDayMapper dayMapper;
    private final ItineraryItemMapper itemMapper;
    private final ItineraryMainMapper mainMapper;
    private final CacheManager cacheManager;
    private final ItineraryEventPublisher eventPublisher;
    private final ObjectMapper objectMapper = JsonMapper.builder().build();

    public ItineraryEnricher(AgentService agentService, ItineraryDayMapper dayMapper,
                             ItineraryItemMapper itemMapper, ItineraryMainMapper mainMapper,
                             CacheManager cacheManager, ItineraryEventPublisher eventPublisher) {
        this.agentService = agentService;
        this.dayMapper = dayMapper;
        this.itemMapper = itemMapper;
        this.mainMapper = mainMapper;
        this.cacheManager = cacheManager;
        this.eventPublisher = eventPublisher;
    }

    /**
     * 富化增强：管家讲解写回 plan_note + 景点介绍逐条写回 intro，结束后精确失效详情缓存。
     * 拒绝（池满）与失败都只影响富化本身，不影响行程主流程。
     */
    @Async("enricherExecutor")
    public void enrichItinerary(ItineraryMain main, GenerateRequest request, Long itineraryId) {
        try {
            List<Map<String, Object>> plans = new ArrayList<>();
            for (ItineraryDay day : dayMapper.selectList(new LambdaQueryWrapper<ItineraryDay>()
                    .eq(ItineraryDay::getItineraryId, itineraryId).orderByAsc(ItineraryDay::getDayNo))) {
                List<String> names = itemMapper.selectList(new LambdaQueryWrapper<ItineraryItem>()
                                .eq(ItineraryItem::getDayId, day.getId()).orderByAsc(ItineraryItem::getSortNo))
                        .stream().map(ItineraryItem::getPoiName).filter(n -> n != null && !n.isBlank()).toList();
                if (names.isEmpty()) {
                    continue;
                }
                plans.add(Map.of("day_no", day.getDayNo(), "items", names));
            }
            if (!plans.isEmpty()) {
                writeButlerNote(main, request, itineraryId, plans);
            }
        } catch (Exception e) {
            log.warn("butler note failed for {}: {}", itineraryId, e.getMessage());
            // 降级通知：讲解失败只影响富化增强，行程主流程不受影响
            eventPublisher.degraded(itineraryId, "butler", e.getMessage(), "跳过讲解");
        }

        // 景点详细介绍：按去重名称分批生成（一次全量在强模型下输出 4000+ token
        // 会被 max_tokens 截断致 JSON 解析失败返回空），逐条单列回填。
        try {
            List<String> names = itemMapper.selectList(new LambdaQueryWrapper<ItineraryItem>()
                            .eq(ItineraryItem::getItineraryId, itineraryId)).stream()
                    .map(ItineraryItem::getPoiName).filter(n -> n != null && !n.isBlank()).distinct().toList();
            if (!names.isEmpty()) {
                Map<String, String> intros = new java.util.HashMap<>();
                final int introBatch = 8;
                for (int i = 0; i < names.size(); i += introBatch) {
                    List<String> chunk = names.subList(i, Math.min(i + introBatch, names.size()));
                    try {
                        AgentPoiIntrosRequest introRequest = new AgentPoiIntrosRequest();
                        introRequest.setCity(main.getCity());
                        introRequest.setNames(chunk);
                        // M3-③：poi-intros 透传意图，与 butlerNote/generate 的 intent 兜底口径一致
                        introRequest.setIntent(ItineraryAsyncPlanner.resolveIntent(request));
                        AgentPoiIntrosResponse introResponse = agentService.poiIntros(introRequest);
                        if (introResponse.getIntros() != null) {
                            intros.putAll(introResponse.getIntros());
                        }
                    } catch (Exception batchEx) {
                        log.warn("poi intros batch failed for {}: {}", itineraryId, batchEx.getMessage());
                    }
                }
                if (!intros.isEmpty()) {
                    // 单列更新而非整行 updateById：富化耗时长，期间用户可能并发编辑同一行，
                    // 整行回写会用陈旧实体覆盖用户改动；且去重后逐名更新减少重复往返。
                    for (String name : names) {
                        String intro = intros.get(name);
                        if (intro != null && !intro.isBlank()) {
                            itemMapper.update(null, new LambdaUpdateWrapper<ItineraryItem>()
                                    .eq(ItineraryItem::getItineraryId, itineraryId)
                                    .eq(ItineraryItem::getPoiName, name)
                                    .set(ItineraryItem::getIntro, intro));
                        }
                    }
                }
            }
        } catch (Exception e) {
            log.warn("poi intros failed for {}: {}", itineraryId, e.getMessage());
            // 降级通知：景点介绍失败同样只影响富化
            eventPublisher.degraded(itineraryId, "poi_intros", e.getMessage(), "跳过景点介绍");
        }

        // 备选池（发现更多）介绍富化：模型单次生成的 60-100 字 intro 遵循度不足，
        // 复用 poi-intros 链路分批补写 200-300 字详细介绍后写回 suggestionsJson。
        // 富化失败只影响备选池介绍文案，不影响主流程。
        try {
            enrichSuggestionIntros(main, request, itineraryId);
        } catch (Exception e) {
            log.warn("suggestion intros failed for {}: {}", itineraryId, e.getMessage());
        }
        evictDetailCache(cacheManager, main.getUserId(), itineraryId);
    }

    /** 读取行程级 suggestionsJson，对缺介绍的点位分批调 poi-intros 补长介绍后单列写回。 */
    private void enrichSuggestionIntros(ItineraryMain main, GenerateRequest request, Long itineraryId) {
        // 富化与主流程异步：重新读一行拿最新 suggestionsJson（避免用 finish 时的陈旧快照）
        ItineraryMain fresh = mainMapper.selectById(itineraryId);
        if (fresh == null) {
            return;
        }
        String json = fresh.getSuggestionsJson();
        if (json == null || json.isBlank()) {
            return;
        }
        List<Map<String, Object>> rows;
        try {
            rows = objectMapper.readValue(json,
                    new tools.jackson.core.type.TypeReference<List<Map<String, Object>>>() {
                    });
        } catch (Exception parseEx) {
            log.warn("suggestions json parse failed for {}: {}", itineraryId, parseEx.getMessage());
            return;
        }
        // 缺介绍的名称按出现顺序分批（poi-intros 按名称数给 max_tokens，批太大易截断）。
        // intro 短于 50 字视为"无有效介绍"（模型单句亮点遵循不了 60-100 字契约），一并富化。
        final int batchSize = 8;
        List<String> pending = rows.stream()
                .filter(r -> {
                    Object intro = r.get("intro");
                    return intro == null || String.valueOf(intro).isBlank()
                            || String.valueOf(intro).length() < 50;
                })
                .map(r -> String.valueOf(r.get("name")))
                .filter(n -> n != null && !n.isBlank() && !"null".equals(n))
                .distinct()
                .toList();
        Map<String, String> intros = new java.util.HashMap<>();
        for (int i = 0; i < pending.size(); i += batchSize) {
            List<String> chunk = pending.subList(i, Math.min(i + batchSize, pending.size()));
            try {
                AgentPoiIntrosRequest introRequest = new AgentPoiIntrosRequest();
                introRequest.setCity(main.getCity());
                introRequest.setNames(chunk);
                introRequest.setIntent(ItineraryAsyncPlanner.resolveIntent(request));
                AgentPoiIntrosResponse resp = agentService.poiIntros(introRequest);
                if (resp.getIntros() != null) {
                    intros.putAll(resp.getIntros());
                }
            } catch (Exception e) {
                log.warn("suggestion intros batch failed for {}: {}", itineraryId, e.getMessage());
            }
        }
        if (intros.isEmpty()) {
            return;
        }
        boolean changed = false;
        for (Map<String, Object> row : rows) {
            Object name = row.get("name");
            if (name == null) {
                continue;
            }
            String intro = intros.get(String.valueOf(name));
            if (intro != null && !intro.isBlank()) {
                row.put("intro", intro);
                changed = true;
            }
        }
        if (changed) {
            try {
                // 单列写回：禁止整行 updateById 覆盖并发编辑
                mainMapper.update(null, new LambdaUpdateWrapper<ItineraryMain>()
                        .eq(ItineraryMain::getId, itineraryId)
                        .set(ItineraryMain::getSuggestionsJson, objectMapper.writeValueAsString(rows)));
            } catch (Exception writeEx) {
                log.warn("suggestions json write failed for {}: {}", itineraryId, writeEx.getMessage());
            }
        }
    }

    private void writeButlerNote(ItineraryMain main, GenerateRequest request, Long itineraryId,
                                 List<Map<String, Object>> plans) throws Exception {
        AgentButlerNoteRequest noteRequest = new AgentButlerNoteRequest();
        noteRequest.setCity(main.getCity());
        noteRequest.setDays(main.getDays());
        noteRequest.setPersons(main.getPersons() == null ? 1 : main.getPersons());
        noteRequest.setPreferences(main.getPreferences() == null ? "" : main.getPreferences());
        noteRequest.setHotelTier(main.getHotelTier() == null ? "" : main.getHotelTier());
        noteRequest.setRegionHint(request.getRegionHint() == null ? "" : request.getRegionHint());
        // Object 字段保持 wire 值与改造前一致：非空为 BigDecimal（JSON 数字），空为 ""
        noteRequest.setBudget(main.getBudget() == null ? "" : main.getBudget());
        noteRequest.setRequirements(request.getRequirements() == null ? "" : request.getRequirements());
        noteRequest.setIntent(ItineraryAsyncPlanner.resolveIntent(request));
        noteRequest.setPlans(plans);
        AgentButlerNoteResponse noteResponse = agentService.butlerNote(noteRequest);
        String note = noteResponse.getNote() == null ? "" : noteResponse.getNote();
        if (!note.isBlank()) {
            // 讲解生成可耗时很久，期间用户可能并发修改同一行程的其它列：
            // 只按 id 精确回写 plan_note 单列，禁止用陈旧实体整行回写，避免丢更新竞态。
            mainMapper.update(null, new LambdaUpdateWrapper<ItineraryMain>()
                    .eq(ItineraryMain::getId, itineraryId)
                    .set(ItineraryMain::getPlanNote, note));
            // 写回成功后通知：preview 截前 60 字，事件体不携带长文本
            eventPublisher.butlerNote(itineraryId, note.length(),
                    note.substring(0, Math.min(60, note.length())));
        }
    }

    /** 落库行程级备选池（发现更多）；失败只记日志，不影响行程主流程。 */
    public void persistSuggestions(Long itineraryId, List<AgentGenerateResponse.Suggestion> suggestions) {
        if (suggestions == null || suggestions.isEmpty()) {
            return;
        }
        try {
            // 单列更新：禁止整行 updateById，避免覆盖用户/富化流程写入的其它字段。
            mainMapper.update(null, new LambdaUpdateWrapper<ItineraryMain>()
                    .eq(ItineraryMain::getId, itineraryId)
                    .set(ItineraryMain::getSuggestionsJson, objectMapper.writeValueAsString(suggestions)));
        } catch (Exception e) {
            log.warn("persist suggestions failed for {}: {}", itineraryId, e.getMessage());
        }
    }

    /**
     * 按 (userId, id) 精确失效详情缓存。key 格式必须与 ItineraryQueryService
     * 的 @Cacheable key "#userId + ':' + #id" 一致；替代原先对整个缓存的全量 clear，
     * 避免一次生成把所有用户的详情缓存全部击穿（J2）。
     */
    static void evictDetailCache(CacheManager cacheManager, Long userId, Long itineraryId) {
        var cache = cacheManager.getCache("itinerary:detail");
        if (cache != null) {
            cache.evict(userId + ":" + itineraryId);
        }
    }
}

File: travel-agent-python/app/services/itinerary_enricher.py
"""生成后的可选富化（移植自 Java `ItineraryEnricher`）。

三段彼此独立的增强，**任何一段失败都只降级、不影响已生成好的行程**：
A. AI 管家讲解 → `itinerary_main.plan_note`
B. 景点详细介绍 → `itinerary_item.intro`（按 8 个一批，避免 max_tokens 截断致整批解析失败）
C. 备选池介绍补写 → 重写 `itinerary_main.suggestions_json`

三条实现纪律（都是并发写回踩过的坑）：
- 一律**单列 UPDATE**，不用陈旧实体整行回写：富化动辄几十秒，期间用户可能正在编辑同一行程；
- B 段按 `poi_name` 批量覆盖同名列（不按 day_id 精确定位），与 Java 一致；
- 收尾无条件按 `(userId, itineraryId)` 精确失效详情缓存（不是全量 clear，避免一次生成
  把所有用户的详情缓存击穿）。
"""

from __future__ import annotations

import json
import logging
from typing import Any

from sqlalchemy import select, update

from app.agent import run_butler_note, run_poi_intros
from app.db.models import ItineraryDay, ItineraryItem, ItineraryMain
from app.db.session import session_scope
from app.schemas.trip import Suggestion
from app.services import itinerary_query

logger = logging.getLogger(__name__)

INTRO_BATCH_SIZE = 8
MIN_SUGGESTION_INTRO = 50
BUTLER_PREVIEW = 60


def persist_suggestions(itinerary_id: int, suggestions: list[Suggestion] | list[dict] | None) -> None:
    """落库行程级备选池（「发现更多」）；失败只记日志。"""
    if not suggestions:
        return
    rows = [
        suggestion.model_dump(mode="json", by_alias=True) if hasattr(suggestion, "model_dump") else dict(suggestion)
        for suggestion in suggestions
    ]
    try:
        payload = json.dumps(rows, ensure_ascii=False)
    except (TypeError, ValueError) as exc:
        logger.warning("persist suggestions failed for %s: %s", itinerary_id, exc)
        return
    with session_scope() as session:
        session.execute(update(ItineraryMain).where(ItineraryMain.id == itinerary_id).values(suggestions_json=payload))


def enrich_itinerary(user_id: int, itinerary_id: int, request: Any) -> None:
    """生成完成后的富化主入口（编排层提交到富化线程池执行）。"""
    with session_scope() as session:
        main = session.get(ItineraryMain, itinerary_id)
    if main is None:
        return
    _write_butler_note(user_id, main, request, itinerary_id)
    _fill_item_intros(user_id, main, request, itinerary_id)
    try:
        _enrich_suggestion_intros(main, request, itinerary_id)
    except Exception as exc:
        logger.warning("suggestion intros failed for %s: %s", itinerary_id, exc)
    itinerary_query.evict_detail(user_id, itinerary_id)


def _plans_for_butler(itinerary_id: int) -> list[dict[str, Any]]:
    """管家讲解的行程摘要：每天 {day_no, items:[点位名]}（键是 snake_case，Python 契约）。"""
    with session_scope() as session:
        days = (
            session.execute(
                select(ItineraryDay).where(ItineraryDay.itinerary_id == itinerary_id).order_by(ItineraryDay.day_no)
            )
            .scalars()
            .all()
        )
        plans: list[dict[str, Any]] = []
        for day in days:
            names = [
                name
                for name in session.execute(
                    select(ItineraryItem.poi_name).where(ItineraryItem.day_id == day.id).order_by(ItineraryItem.sort_no)
                )
                .scalars()
                .all()
                if name and name.strip()
            ]
            if names:
                plans.append({"day_no": day.day_no, "items": names})
    return plans


def _write_butler_note(user_id: int, main: ItineraryMain, request: Any, itinerary_id: int) -> None:
    plans = _plans_for_butler(itinerary_id)
    if not plans:
        return
    try:
        payload = {
            "city": main.city,
            "days": main.days,
            "persons": main.persons if main.persons is not None else 1,
            # 偏好列本身是逗号拼接串；Java 也是原样传字符串，
            # `run_butler_note` 内部 `"、".join(str)` 会逐字符连接——这是迁移前就有的形状，
            # 不在迁移里改提示词口径（会动到评测基线）。
            "preferences": main.preferences or "",
            "hotel_tier": main.hotel_tier or "",
            "region_hint": getattr(request, "region_hint", None) or "",
            "budget": "" if main.budget is None else main.budget,
            "requirements": getattr(request, "requirements", None) or "",
            "intent": resolve_intent(request),
            "plans": plans,
            "validation_log": None,
        }
        note = run_butler_note(payload) or ""
        if not note.strip():
            return
        with session_scope() as session:
            session.execute(update(ItineraryMain).where(ItineraryMain.id == itinerary_id).values(plan_note=note))
        from app.services import generation_events

        generation_events.butler_note(itinerary_id, len(note), note[:BUTLER_PREVIEW])
    except Exception as exc:
        logger.warning("butler note failed for %s: %s", itinerary_id, exc)
        _publish_degraded(itinerary_id, "butler", str(exc), "跳过讲解")


def _fill_item_intros(user_id: int, main: ItineraryMain, request: Any, itinerary_id: int) -> None:
    with session_scope() as session:
        names = list(
            dict.fromkeys(
                name
                for name in session.execute(
                    select(ItineraryItem.poi_name).where(ItineraryItem.itinerary_id == itinerary_id)
                )
                .scalars()
                .all()
                if name and name.strip()
            )
        )
    intros: dict[str, str] = {}
    try:
        for start in range(0, len(names), INTRO_BATCH_SIZE):
            chunk = names[start : start + INTRO_BATCH_SIZE]
            try:
                intros.update(run_poi_intros(main.city, chunk, resolve_intent(request)) or {})
            except Exception as exc:
                logger.warning("poi intros batch failed for %s: %s", itinerary_id, exc)
        with session_scope() as session:
            for name, intro in intros.items():
                if intro and intro.strip():
                    session.execute(
                        update(ItineraryItem)
                        .where(ItineraryItem.itinerary_id == itinerary_id, ItineraryItem.poi_name == name)
                        .values(intro=intro)
                    )
    except Exception as exc:
        logger.warning("poi intros failed for %s: %s", itinerary_id, exc)
        _publish_degraded(itinerary_id, "poi_intros", str(exc), "跳过景点介绍")


def _enrich_suggestion_intros(main: ItineraryMain, request: Any, itinerary_id: int) -> None:
    with session_scope() as session:
        fresh = session.get(ItineraryMain, itinerary_id)  # 用最新行，避免拿 finish 时的陈旧快照
        raw = fresh.suggestions_json if fresh is not None else None
    if not raw or not raw.strip():
        return
    try:
        rows = json.loads(raw)
    except (ValueError, TypeError) as exc:
        logger.warning("suggestions json parse failed for %s: %s", itinerary_id, exc)
        return
    if not isinstance(rows, list) or not rows:
        return

    def needs_intro(row: dict) -> bool:
        intro = row.get("intro")
        return intro is None or not str(intro).strip() or len(str(intro)) < MIN_SUGGESTION_INTRO

    # 只补「没有有效介绍」的行：模型单次生成的 60-100 字亮点遵循度不足，短于 50 字视为无介绍
    names = list(
        dict.fromkeys(
            str(row.get("name"))
            for row in rows
            if isinstance(row, dict)
            and needs_intro(row)
            and row.get("name") is not None
            and str(row.get("name")).strip()
            and str(row.get("name")) != "null"
        )
    )

    intros: dict[str, str] = {}
    for start in range(0, len(names), INTRO_BATCH_SIZE):
        try:
            intros.update(
                run_poi_intros(main.city, names[start : start + INTRO_BATCH_SIZE], resolve_intent(request)) or {}
            )
        except Exception as exc:
            logger.warning("suggestion intros batch failed for %s: %s", itinerary_id, exc)
    if not intros:
        return

    changed = False
    for row in rows:
        if not isinstance(row, dict):
            continue
        name = row.get("name")
        intro = intros.get(str(name)) if name is not None else None
        if intro and intro.strip():
            row["intro"] = intro
            changed = True
    if not changed:
        return
    try:
        with session_scope() as session:
            session.execute(
                update(ItineraryMain)
                .where(ItineraryMain.id == itinerary_id)
                .values(suggestions_json=json.dumps(rows, ensure_ascii=False))
            )
    except (TypeError, ValueError) as exc:
        logger.warning("suggestions json write failed for %s: %s", itinerary_id, exc)


def resolve_intent(request: Any) -> str:
    """意图兜底：未填 intent 时用 requirements 作为意图（M1 契约）。"""
    if request is None:
        return ""
    intent = getattr(request, "intent", None)
    if intent and intent.strip():
        return intent.strip()
    return getattr(request, "requirements", None) or ""


def _publish_degraded(itinerary_id: int, scope: str, reason: str, fallback: str) -> None:
    # 延迟导入：富化 → 事件 → trace，避免模块级环依赖
    from app.services import generation_events

    generation_events.degraded(itinerary_id, scope, reason, fallback)
